core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroup.java

```diff
@@ -16,150 +16,183 @@
-     */
     @Override
-    public void close() {
-        if (scheduledFuture != null) {
-            scheduledFuture.cancel(true);
+    ImmutableSortedSet<Endpoint> onDnsRecords(List<DnsRecord> records, int ttl) throws Exception {
+        final ImmutableSortedSet.Builder<Endpoint> builder = ImmutableSortedSet.naturalOrder();
+        final boolean hasLoopbackARecords =
+                records.stream()
+                       .filter(r -> r instanceof DnsRawRecord)
+                       .map(r -> (DnsRawRecord) r)
+                       .anyMatch(r -> r.type() == DnsRecordType.A &&
+                                     r.content().getByte(r.content().readerIndex()) == 127);
+
+        for (DnsRecord r : records) {
+            if (!(r instanceof DnsRawRecord)) {
+                continue;
+            }
+
+            final DnsRecordType type = r.type();
+            final ByteBuf content = ((ByteBufHolder) r).content();
+            final int contentLen = content.readableBytes();
+
+            // Skip invalid records.
+            if (type == DnsRecordType.A) {
+                if (contentLen != 4) {
+                    warnInvalidRecord(DnsRecordType.A, content);
+                    continue;
+                }
+            } else if (type == DnsRecordType.AAAA) {
+                if (contentLen != 16) {
+                    warnInvalidRecord(DnsRecordType.AAAA, content);
+                    continue;
+                }
+            } else {
+                continue;
+            }
+
+            // Convert the content into an IP address into an endpoint.
+            final String ipAddr;
+            final byte[] addrBytes = new byte[contentLen];
+            content.getBytes(content.readerIndex(), addrBytes);
+
+            if (contentLen == 16) {
+                // 'result' contains both A and AAAA records.
+                // Convert some IPv6 addresses into IPv4 addresses to remove duplicate endpoints.
+                if (addrBytes[0] == 0x00 && addrBytes[1] == 0x00 &&
+                    addrBytes[2] == 0x00 && addrBytes[3] == 0x00 &&
+                    addrBytes[4] == 0x00 && addrBytes[5] == 0x00 &&
+                    addrBytes[6] == 0x00 && addrBytes[7] == 0x00 &&
+                    addrBytes[8] == 0x00 && addrBytes[9] == 0x00) {
+
+                    if (addrBytes[10] == 0x00 && addrBytes[11] == 0x00) {
+                        if (addrBytes[12] == 0x00 && addrBytes[13] == 0x00 &&
+                            addrBytes[14] == 0x00 && addrBytes[15] == 0x01) {
+                            // Loopback address (::1)
+                            if (hasLoopbackARecords) {
+                                // Contains an IPv4 loopback address already; skip.
+                                continue;
+                            } else {
+                                ipAddr = "::1";
+                            }
+                        } else {
+                            // IPv4-compatible address.
```

Comment:
Out of curiosity, do you think Netty should handle this conversion? I found this document about compatibility in Oracle docs and wonder if the JDK does this transparently.

https://docs.oracle.com/cd/E19683-01/817-0573/transition-4/index.html